src/compiler-src.md

```diff
@@ -1,137 +1,191 @@
-
-The dependency structure of these crates is roughly a diamond:
-
-```text
-                  rustc_driver
-                /      |       \
-              /        |         \
-            /          |           \
-          /            v             \
-rustc_codegen  rustc_borrowck   ...  rustc_metadata
-          \            |            /
-            \          |          /
-              \        |        /
-                \      v      /
-                  rustc_middle
-                       |
-                       v
-                   rustc_ast
-                    /    \
-                  /       \
-           rustc_span  rustc_builtin_macros
-```
-
-The `rustc_driver` crate, at the top of this lattice, is effectively
-the "main" function for the rust compiler. It doesn't have much "real
-code", but instead ties together all of the code defined in the other
-crates and defines the overall flow of execution. (As we transition
-more and more to the [query model], however, the
-"flow" of compilation is becoming less centrally defined.)
-
-At the other extreme, the `rustc_middle` crate defines the common and
-pervasive data structures that all the rest of the compiler uses
-(e.g. how to represent types, traits, and the program itself). It
-also contains some amount of the compiler itself, although that is
-relatively limited.
-
-Finally, all the crates in the bulge in the middle define the bulk of
-the compiler – they all depend on `rustc_middle`, so that they can make use
-of the various types defined there, and they export public routines
-that `rustc_driver` will invoke as needed (more and more, what these
-crates export are "query definitions", but those are covered later
-on).
-
-Below `rustc_middle` lie various crates that make up the parser and error
-reporting mechanism. They are also an internal part
-of the compiler and not intended to be stable (though they do wind up
-getting used by some crates in the wild; a practice we hope to
-gradually phase out).
-
-## The main stages of compilation
-
-The Rust compiler is in a bit of transition right now. It used to be a
-purely "pass-based" compiler, where we ran a number of passes over the
-entire program, and each did a particular check of transformation. We
-are gradually replacing this pass-based code with an alternative setup
-based on on-demand **queries**. In the query-model, we work backwards,
-executing a *query* that expresses our ultimate goal (e.g. "compile
-this crate"). This query in turn may make other queries (e.g. "get me
-a list of all modules in the crate"). Those queries make other queries
-that ultimately bottom out in the base operations, like parsing the
-input, running the type-checker, and so forth. This on-demand model
-permits us to do exciting things like only do the minimal amount of
-work needed to type-check a single function. It also helps with
-incremental compilation. (For details on defining queries, check out
-the [query model].)
```

Comment:
I think we should keep this paragraph.

Reply:
Also, if we need examples of _exciting_ things with queries, I've been doing lots of fancy things for https://github.com/rust-lang/rust/pull/73566 (e.g. [not running typeck at all](https://github.com/rust-lang/rust/pull/73566/files#diff-0de644786a2f1fd88c7d7f44bc3fa4bbR387)).

Reply:
@jyn514 I feel like this paragraph is redundant to a few other places in the guide (e.g. the overview chapter or the queries chapter), whereas this chapter is more about the structure of the code itself.

Regarding the example, I think it could work well in one of the above mentioned chapters.

Reply:
Ah good point, I forgot that there were two whole chapters just for queries. In that case removing this seems fine.